llarp/nodedb.cpp

```diff
@@ -112,6 +114,89 @@ llarp_nodedb::FindClosestTo(const llarp::dht::Key_t &location)
   return rc;
 }
 
+std::vector< llarp::RouterContact >
+llarp_nodedb::FindClosestTo(const llarp::dht::Key_t &location,
+                            uint32_t numRouters)
+{
+  if(numRouters < num_loaded())
+  {
+    numRouters = num_loaded();
+  }
+  /*
+   * XXX: this attempts to use std::partial_sort_copy(), which has a few
+  requirements
+   * we can't quite meet:
+   *
+   * 1) the iterators must operate on same types (NetDBMap_t would give us
+  pairs)
+   * 2) the second set of iterators must be of types which:
+   *     - swap is defined
+   *     - are move-constructible
+   *     - are move-assignable
+   *    std::pair is none (or at least not all) of these things
+   * 3) the result iterators must specify a range; so their container must not
+  be empty
+   *
+
+  using NetDBPair_t = std::pair< llarp::RouterID, NetDBEntry >;
+
+  struct Compare
+  {
+    const llarp::dht::Key_t target;
+
+    Compare(const llarp::dht::Key_t& target_) : target(target_)
+    {
+    }
+
+    bool
+    operator()(const NetDBPair_t& left, const NetDBPair_t& right) const
+    {
+      return (left.second.rc.pubkey ^ target) < (right.second.rc.pubkey ^
+  target);
+    }
+  };
+
+  llarp::util::Lock lock(&access);
+
+  std::vector< NetDBPair_t > closestEntries;
+  closestEntries.resize(numRouters); // so that we have a valid iterator range
+
+  const Compare compare(location);
+
+  std::partial_sort_copy(
+      entries.begin(), entries.end(),
+      closestEntries.begin(), closestEntries.end(),
+      compare);
+
+  std::vector< llarp::RouterContact > closest;
+  closest.reserve(closestEntries.size());
+
+  for (const auto& entry : closestEntries) {
+    closest.push_back(entry.second.rc);
+  }
+
+  return closest;
+  */
+
+  // TODO: avoid this ugly copy
+  std::vector< llarp::RouterContact > all;
+  all.reserve(entries.size());
+  for(auto &entry : entries)
+  {
+    all.push_back(entry.second.rc);
+  }
+
+  std::vector< llarp::RouterContact > closest;
+  closest.resize(numRouters);
```

Comment:
Perhaps instead of storing all the RouterContacts, just store pointers to the `decltype(entries)::value_type*`, then do a `std::partial_sort` on that with a custom sort function that compares distance (of the deferenced pointer), then finally deref and copy those sorted elements into full `RouterContacts` in `closest`.

Reply:
alternatively store the keys in that vector instead and then fetch the record via key after.

Reply:
Good thoughts, either solution is probably an adequate stop-gap.

Reply:
This is just an obnoxious problem -- there are a lot of inflexible pieces at play here (`XorMetric`, `AlignedBuffer` and its inheritance tree, `partial_sort_copy()`,  `NetDBMap_t`, etc).

I suggest:
1) we not let this hold up general testing
2) in parallel, work on a hand-rolled algorithm OR
2b) just specialize the function to look for exactly 2 or 4 routers

Reply:
This is rewritten now to sort pointers.